# Data Understanding After Cleaning

After performing the cleaning, we check again some basic information to show that all the data in correct in the sintax and the semantic

In [1]:
# Autoreload allows the notebook to dynamically load code: if we update some helper functions *outside* of the notebook, we do not need to reload the notebook.
%load_ext autoreload
%autoreload 2

In [2]:
import pandas as pd
import numpy as np
import re
import unicodedata

In [3]:
csv_file = "../data/merged_dataset.csv"
dataset = pd.read_csv(csv_file)
dataset.head()

cyclist_url         cyclist_name  birth_year  weight  height  \
0       gerard-vianen       Gerard  Vianen      1944.0     NaN     NaN   
1        jos-huysmans        Jos  Huysmans      1941.0     NaN     NaN   
2     romain-furniere     Romain  Furniere      1943.0     NaN     NaN   
3         willy-donie         Willy  Donie      1945.0     NaN     NaN   
4  frans-melckenbeeck  Frans  Melckenbeeck      1940.0     NaN     NaN   

   nationality                      race_url         race_name  points  \
0  Netherlands  amstel-gold-race/1970/result  Amstel Gold Race   225.0   
1      Belgium  amstel-gold-race/1970/result  Amstel Gold Race   225.0   
2      Belgium  amstel-gold-race/1970/result  Amstel Gold Race   225.0   
3      Belgium  amstel-gold-race/1970/result  Amstel Gold Race   225.0   
4      Belgium  amstel-gold-race/1970/result  Amstel Gold Race   225.0   

   uci_points  ...  profile  startlist_quality        date  position  \
0         NaN  ...      2.0                372  1970-04-25        23   
1         NaN  ...      2.0                372  1970-04-25        18   
2         NaN  ...      2.0                372  1970-04-25        31   
3         NaN  ...      2.0                372  1970-04-25        27   
4         NaN  ...      2.0                372  1970-04-25         8   

              cyclist  cyclist_age mostly_tarmac             cyclist_team  \
0       gerard-vianen         26.0         False   uae-team-emirates-2018   
1        jos-huysmans         29.0         False          team-monex-2005   
2     romain-furniere         27.0         False                      NaN   
3         willy-donie         25.0         False                      NaN   
4  frans-melckenbeeck         30.0         False  team-qhubeka-assos-2021   

   delta race_stage  
0   52.0     result  
1   52.0     result  
2   52.0     result  
3   52.0     result  
4   33.0     result  

[5 rows x 22 columns]

In [4]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 589739 entries, 0 to 589738
Data columns (total 22 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   cyclist_url        589739 non-null  object 
 1   cyclist_name       589739 non-null  object 
 2   birth_year         589739 non-null  float64
 3   weight             480007 non-null  float64
 4   height             480007 non-null  float64
 5   nationality        589739 non-null  object 
 6   race_url           589739 non-null  object 
 7   race_name          589739 non-null  object 
 8   points             589739 non-null  float64
 9   uci_points         422184 non-null  float64
 10  length             589739 non-null  float64
 11  climb_total        589739 non-null  float64
 12  profile            589739 non-null  float64
 13  startlist_quality  589739 non-null  int64  
 14  date               589739 non-null  object 
 15  position           589739 non-null  int64  
 16  cy

In [5]:
dataset.describe()

birth_year         weight         height         points  \
count  589739.000000  480007.000000  480007.000000  589739.000000   
mean     1977.683221      68.362072     179.699850      89.225897   
std        11.537102       6.182201       6.269585      54.415806   
min      1933.000000      48.000000     154.000000      18.000000   
25%      1970.000000      64.000000     175.000000      50.000000   
50%      1979.000000      68.000000     180.000000      80.000000   
75%      1986.000000      73.000000     184.000000     100.000000   
max      2004.000000      94.000000     204.000000     350.000000   

          uci_points         length    climb_total        profile  \
count  422184.000000  589739.000000  589739.000000  589739.000000   
mean       72.048275  166776.290800    2320.190052       2.483317   
std       121.041689   64543.334848    1334.129873       1.459891   
min         6.000000    1000.000000       2.000000       1.000000   
25%        16.000000  152500.000000    1355.000000       1.000000   
50%        20.000000  178200.000000    2312.000000       2.000000   
75%       100.000000  203500.000000    3217.000000       4.000000   
max       800.000000  338000.000000    6974.000000       5.000000   

       startlist_quality       position    cyclist_age          delta  
count      589739.000000  589739.000000  589739.000000  589739.000000  
mean         1101.182822      74.217408      28.485834     430.755887  
std           380.588233      48.405261       3.855921    1011.486288  
min           115.000000       0.000000      13.000000       0.000000  
25%           844.000000      32.000000      26.000000      11.000000  
50%           988.000000      70.000000      28.000000     158.000000  
75%          1309.000000     112.000000      31.000000     626.000000  
max          2047.000000     209.000000      56.000000   46380.000000

## Checks of cyclists data

Now we start from the data related to the initial cyclist dataset

### Check on 'cyclist_url' data

We start considering the `cyclys_url` column, and check the number of null values and the count the occurrences of each unique value

In [8]:
print('Total number of null values in cyclist_url column: ' + str(dataset['cyclist_url'].isnull().sum())
      + ' (' + str(round(dataset['cyclist_url'].isnull().sum() / len(dataset) * 100, 2)) + '%)')

print('\nCount occurrences of each value in _url column:')
url_counts = dataset['cyclist_url'].value_counts()
print(url_counts)

Total number of null values in cyclist_url column: 0 (0.0%)

Count occurrences of each value in _url column:
cyclist_url
matteo-tosatto        959
alejandro-valverde    942
luis-leon-sanchez     899
imanol-erviti         883
haimar-zubeldia       883
                     ... 
wim-dubois              1
james-lewis-perry       1
marcio-may              1
john-lieswyn            1
romain-furniere         1
Name: count, Length: 6095, dtype: int64


### Check on 'cyclist_name' data

Now we consider the `cyclist_name` column, and check the number of null values and the count the occurrences of each unique value

In [9]:
print('Total number of null values in cyclist_name column: ' + str(dataset['cyclist_name'].isnull().sum())
        + ' (' + str(round(dataset['cyclist_name'].isnull().sum() / len(dataset) * 100, 2)) + '%)')

print('\nCount occurrences of each value in name column:')
name_counts = dataset['cyclist_name'].value_counts()
print(name_counts)   

Total number of null values in cyclist_name column: 0 (0.0%)

Count occurrences of each value in name column:
cyclist_name
Matteo  Tosatto                   959
Alejandro  Valverde               942
Luis León  Sánchez                899
Imanol  Erviti                    883
Haimar  Zubeldia                  883
                                 ... 
John  Lieswyn                       1
Makoto  Iijima                      1
Bruno Ricardo Fernandes  Neves      1
Anatoliy  Budyak                    1
Miyataka  Shimizu                   1
Name: count, Length: 6088, dtype: int64


In [10]:
# Check if the 'cyclist_name' column contains any number
invalid_rows = dataset[dataset['cyclist_name'].str.contains(r'\d')]

print('Number of rows with numbers in the name: ' + str(len(invalid_rows)))

Number of rows with numbers in the name: 0


### Check on 'birth_year' data

Now we consider the `birth_year` column, and check the number of null values and the count the occurrences of each unique value

In [11]:
print('Total number of null values in birth_year column: ' + str(dataset['birth_year'].isnull().sum())
        + ' (' + str(round(dataset['birth_year'].isnull().sum() / len(dataset) * 100, 2)) + '%)')

print('\nCount occurrences of each value in birth_year column:')
birth_year_counts = dataset['birth_year'].value_counts()
print(birth_year_counts)

Total number of null values in birth_year column: 0 (0.0%)

Count occurrences of each value in birth_year column:
birth_year
1980.0    23848
1985.0    23585
1986.0    21746
1974.0    21021
1981.0    18993
          ...  
2004.0       25
1938.0       13
1937.0       13
1934.0        5
1933.0        1
Name: count, Length: 71, dtype: int64


In [13]:
# Remove '.0' from 'birth_year' and check if it's a valid year in the form '19nn' or '20nn'
def is_valid_year(birth_year):
    # Convert to string and remove '.0'
    year = str(birth_year).replace('.0', '')
    # Check if the year is a digit and starts with '19' or '20'
    return year.isdigit() and (year.startswith('19') or year.startswith('20'))

# Filter out rows where 'birth_year' is not null
invalid_rows = dataset[dataset['birth_year'].notnull()]

# Apply the validation function across the 'birth_year' column
invalid_rows = invalid_rows[~invalid_rows['birth_year'].apply(is_valid_year)]

print('Number of invalid birth_year: ' + str(len(invalid_rows)))


Number of invalid birth_year: 0


### Check on 'weight' data

Now we consider the `weight` column, and check the number of null values and the count the occurrences of each unique value

In [15]:
print('Total number of null values in weight column: ' + str(dataset['weight'].isnull().sum())
        + ' (' + str(round(dataset['weight'].isnull().sum() / len(dataset) * 100, 2)) + '%)')

print('\nCount occurrences of each value in weight column:')
weight_counts = dataset['weight'].value_counts()
print(weight_counts)

Total number of null values in weight column: 109732 (18.61%)

Count occurrences of each value in weight column:
weight
70.0    41266
65.0    37681
68.0    34220
72.0    26398
67.0    26346
        ...  
67.5       23
79.5       22
68.8       20
74.5       10
73.5        7
Name: count, Length: 69, dtype: int64


### Check on 'height' data

Now we consider the `height` column, and check the number of null values and the count the occurrences of each unique value

In [16]:
print('Total number of null values in height column: ' + str(dataset['height'].isnull().sum())
        + ' (' + str(round(dataset['height'].isnull().sum() / len(dataset) * 100, 2)) + '%)')

print('\nCount occurrences of each value in height column:')
height_counts = dataset['height'].value_counts()
print(height_counts)

Total number of null values in height column: 109732 (18.61%)

Count occurrences of each value in height column:
height
180.0    38280
178.0    35610
181.0    31417
183.0    29743
175.0    26796
182.0    25377
179.0    23291
185.0    23019
173.0    21466
174.0    21433
176.0    21274
184.0    21202
186.0    20068
177.0    19551
190.0    16416
172.0    14020
170.0    11917
171.0    11215
187.0    10760
188.0     9676
169.0     8617
191.0     7677
189.0     6492
164.0     3501
193.0     3061
167.0     2824
168.0     2473
192.0     2439
194.0     2320
165.0     2107
195.0     1243
197.0     1182
198.0     1012
159.0      558
166.0      511
199.0      365
160.0      220
196.0      189
162.0      187
154.0      185
161.0      111
202.0       61
204.0       55
157.0       33
163.0       21
200.0       21
158.0        8
155.0        3
Name: count, dtype: int64


### Check on 'nationality' data

Now we consider the `nationality` column, and check the number of null values and the count the occurrences of each unique value

In [17]:
print('Total number of null values in nationality column: ' + str(dataset['nationality'].isnull().sum())
        + ' (' + str(round(dataset['nationality'].isnull().sum() / len(dataset) * 100, 2)) + '%)')

print('\nCount occurrences of each value in nationality column:')
nationality_counts = dataset['nationality'].value_counts()
print(nationality_counts)

Total number of null values in nationality column: 0 (0.0%)

Count occurrences of each value in nationality column:
nationality
Italy            102130
Spain             91041
France            84916
Belgium           54343
Netherlands       36385
                  ...  
Taiwan                3
Liechtenstein         1
Tunisia               1
Azerbaijan            1
Hongkong              1
Name: count, Length: 72, dtype: int64


## Checks of races data

### Check on 'race_url' data

We start considering the `_url` column, and check the number of null values and the count the occurrences of each unique value

In [ ]:
print('Total number of null values in _url column: ' + str(dataset['_url'].isnull().sum())
      + ' (' + str(round(dataset['_url'].isnull().sum() / len(dataset) * 100, 2)) + '%)')

print('\nCount occurrences of each value in _url column:')
url_counts = dataset['_url'].value_counts()
print(url_counts)

Total number of null values in _url column: 0 (0.0%)

Count occurrences of each value in _url column:
_url
tour-de-france/1986/stage-1     210
tour-de-france/1986/prologue    210
tour-de-france/1987/prologue    207
giro-d-italia/2011/stage-2      206
tour-de-france/1987/stage-1     206
                               ... 
il-lombardia/1972/result          1
giro-d-italia/2001/stage-14       1
tour-de-suisse/1972/stage-2       1
tour-de-suisse/1972/stage-7       1
tour-de-suisse/1972/stage-6b      1
Name: count, Length: 5281, dtype: int64


We have lots of different values, but no null values

In this block we check if there are `_url` values that are not in the form name/year/stage

In [ ]:
# Split url by / in name, year and stage
url_split = dataset['_url'].str.split('/', expand=True)
# Check null elements in url_split[0], url_split[1] and url_split[2], and if url_split[1] contains only digits
invalid_rows = dataset[url_split[0].isnull() | url_split[1].isnull() | url_split[2].isnull() | ~url_split[1].str.isdigit()]
print('Number of invalid URLs: ' + str(len(invalid_rows)))

Number of invalid URLs: 0


In [ ]:
#TODO: Chiedere sulle gare dove ci sono un solo o pochi partecipanti

url_counts = dataset['_url'].value_counts()

# From url_counts, get the urls where the number of occurrence is less than 2
print(url_counts[url_counts < 6])

_url
tour-de-romandie/1992/stage-6     5
dauphine/1983/prologue            5
tour-de-romandie/1989/stage-1     5
tirreno-adriatico/1996/stage-7    5
volta-a-catalunya/1992/stage-2    5
                                 ..
il-lombardia/1972/result          1
giro-d-italia/2001/stage-14       1
tour-de-suisse/1972/stage-2       1
tour-de-suisse/1972/stage-7       1
tour-de-suisse/1972/stage-6b      1
Name: count, Length: 146, dtype: int64


### Check on 'race_name' data

Now we consider the `name` column, and check the number of null values and the count the occurrences of each unique value

In [ ]:
print('Total number of null values in name column: ' + str(dataset['name'].isnull().sum())
      + ' (' + str(round(dataset['name'].isnull().sum() / len(dataset) * 100, 2)) + '%)')

print('\n\nCount occurrences of each value in name column:')
name_counts = dataset['name'].value_counts()
print(name_counts)

Total number of null values in name column: 0 (0.0%)


Count occurrences of each value in name column:
name
Tour de France                     145500
Giro d'Italia                       95581
Vuelta a España                     89222
Tour de Suisse                      33682
Paris - Nice                        32362
                                    ...  
E3 Saxo Classic                       101
E3 BinckBank Classic                   99
E3 Prijs Vlaanderen - Harelbeke        98
Clásica Ciclista San Sebastian         84
Clásica San Sebastián                  52
Name: count, Length: 61, dtype: int64


We have different values, but no null values

Since it's small, we print all the values

In [ ]:
# Print all the names that appear, alphabetically ordered
print(name_counts.index.sort_values())

Index(['Amstel Gold Race', 'Clasica Ciclista San Sebastian',
       'Clásica Ciclista San Sebastian', 'Clásica Ciclista San Sebastián',
       'Clásica San Sebastián', 'Criterium du Dauphiné',
       'Criterium du Dauphiné Libére', 'Critérium du Dauphiné',
       'Critérium du Dauphiné Libéré', 'Donostia San Sebastian Klasikoa',
       'Dwars door België / À travers la Belgique', 'Dwars door Vlaanderen',
       'Dwars door Vlaanderen - A travers la Flandre ME',
       'Dwars door Vlaanderen / A travers la Flandre',
       'Dwars door Vlaanderen / A travers la Flandre ME',
       'E3 BinckBank Classic', 'E3 Harelbeke', 'E3 Prijs Vlaanderen',
       'E3 Prijs Vlaanderen - Harelbeke', 'E3 Saxo Bank Classic',
       'E3 Saxo Classic', 'E3-Prijs Harelbeke', 'Giro d'Italia',
       'Giro di Lombardia', 'Gran Camiño', 'Grand Prix Cycliste de Montréal',
       'Grand Prix Cycliste de Quebec', 'Grand Prix Cycliste de Québec',
       'Il Lombardia', 'Itzulia Basque Country', 'La Flèche Wallonne'

For each url, check if all the `name` values are the same


In [ ]:
# Group by '_url' and calculate the number of unique values in the 'name' column
name_uniques = dataset.groupby('_url')['name'].nunique(dropna=False)
# Filter the URLs with more than one unique name
multiple_names_urls = name_uniques[name_uniques > 1].index

print('Number of URLs with more than one unique name: ' + str(len(multiple_names_urls)))

Number of URLs with more than one unique name: 0


In this block we are checking if there are `name` values that contains any incorrect numbers

In [ ]:
# Get rows wehere 'name' value contains any number, except for names containing 'E3' (there are some races with E3 in the name)
invalid_rows = dataset[dataset['name'].str.contains(r'\d') & ~dataset['name'].str.contains('E3')]
print('Number of invalid names: ' + str(len(invalid_rows)))

Number of invalid names: 0


### Check on 'points' data

Now we consider the `point` column, and check the number of null values and the count the occurrences of each unique value

In [ ]:
print('Total number of null values in points column: ' + str(dataset['points'].isnull().sum())
      + ' (' + str(round(dataset['points'].isnull().sum() / len(dataset) * 100, 2)) + '%)')

print('\nCount occurrences of each value in points column:')
point_counts = dataset['points'].value_counts()
print(point_counts)

Total number of null values in points column: 477 (0.08%)

Count occurrences of each value in points column:
points
80.0     198878
50.0     186102
100.0    141706
275.0     22299
225.0     19536
125.0      5992
30.0       4313
350.0      3917
70.0       3299
75.0       1963
20.0        792
18.0        292
35.0        183
150.0       116
Name: count, dtype: int64


We have not a lot of values, and few null values. Also, we see that every value is sintatically correct

For each url, check if all the `name` values are the same

In [ ]:
# Group by '_url' and calculate the number of unique values in the 'points' column
points_uniques = dataset.groupby('_url')['points'].nunique(dropna=False)
# Filter the URLs with more than one unique points
multiple_points_urls = points_uniques[points_uniques > 1].index

print('Number of URLs with more than one unique points: ' + str(len(multiple_points_urls)))

Number of URLs with more than one unique points: 0


We check the urls where `points` is null

In [ ]:
# Get unique data based on '_url' and 'points'
unique_data = dataset.drop_duplicates(subset=['_url', 'points'])
# Get rows where 'points' is null
rows = unique_data[unique_data['points'].isnull()]

print('Number of urls with null points: ' + str(len(rows)))
print('\nURLs of the rows with null points:')
print(rows['_url'])

Number of urls with null points: 4

URLs of the rows with null points:
156755    vuelta-a-espana/1994/stage-5
461300    tour-de-france/1986/stage-19
517517    tour-de-france/1988/prologue
561313    tour-de-france/2019/stage-19
Name: _url, dtype: object


### Check on 'uci_points' data

Now we consider the `uci_points` column, and check the number of null values and the count the occurrences of each unique value

In [ ]:
print('Total number of null values in uci_points column: ' + str(dataset['uci_points'].isnull().sum())
      + ' (' + str(round(dataset['uci_points'].isnull().sum() / len(dataset) * 100, 2)) + '%)')

print('\nCount occurrences of each value in uci_points column:')
uci_point_counts = dataset['uci_points'].value_counts()
print(uci_point_counts)

Total number of null values in uci_points column: 338779 (57.43%)

Count occurrences of each value in uci_points column:
uci_points
100.0    47640
6.0      43390
16.0     41103
60.0     39317
20.0     21303
120.0    20785
50.0     13266
500.0     6096
40.0      4102
210.0     3449
180.0     3103
400.0     2571
300.0     2057
14.0       792
600.0      675
800.0      514
200.0      338
80.0       328
10.0       148
402.0      109
Name: count, dtype: int64


We have different values, but a lot of null values. Also, we see that every value is sintatically correct

For each url, check if all the `uci_points` values are the same

In [ ]:
# Group by '_url' and calculate the number of unique values in the 'uci_points' column
uci_points_uniques = dataset.groupby('_url')['uci_points'].nunique(dropna=False)
# Filter the URLs with more than one unique uci_points
multiple_uci_points_urls = uci_points_uniques[uci_points_uniques > 1].index

print('Number of URLs with more than one unique uci_points: ' + str(len(multiple_uci_points_urls)))

Number of URLs with more than one unique uci_points: 0


We check the urls where `uci_points` is null

In [ ]:
# Get unique data based on '_url' and 'uci_points'
unique_data = dataset.drop_duplicates(subset=['_url', 'uci_points'])
# Get rows where 'uci_points' is null
rows = unique_data[unique_data['uci_points'].isnull()]

print('Number of rows with null uci_points: ' + str(len(rows)))
print('\nURLs of the rows with null uci_points:')
print(rows['_url'])

Number of rows with null uci_points: 3682

URLs of the rows with null uci_points:
0             tour-de-france/1978/stage-6
426       volta-a-catalunya/1999/prologue
866          tour-de-france/1978/stage-14
1075      volta-a-catalunya/1981/stage-2b
1084             paris-nice/1994/stage-8b
                       ...               
588250              gp-quebec/2010/result
588712       tour-de-france/1982/stage-21
588837     tirreno-adriatico/1993/stage-3
589397          paris-roubaix/2000/result
589463            paris-nice/1976/stage-2
Name: _url, Length: 3682, dtype: object


We check rows where `point` is null but not `uci_point`

In [ ]:
# Get unique data based on '_url' and 'uci_points'
unique_data = dataset.drop_duplicates(subset=['_url', 'uci_points'])
# Get rows where 'uci_points' is not null, but 'points' is null
rows = unique_data[unique_data['uci_points'].notnull() & unique_data['points'].isnull()]

print('Number of rows with null uci_points: ' + str(len(rows)))

Number of rows with null uci_points: 0


### Check on 'length' data

Now we consider the `length` column, and check the number of null values and the count the occurrences of each unique value

In [ ]:
print('Total number of null values in length column: ' + str(dataset['length'].isnull().sum())
      + ' (' + str(round(dataset['length'].isnull().sum() / len(dataset) * 100, 2)) + '%)')

print('\nCount occurrences of each value in length column:')
length_counts = dataset['length'].value_counts()
print(length_counts)

Total number of null values in length column: 0 (0.0%)

Count occurrences of each value in length column:
length
177000.0    5039
170000.0    4717
195000.0    4413
200000.0    4401
178000.0    4286
            ... 
4500.0         5
107000.0       5
123100.0       5
142700.0       5
2000.0         3
Name: count, Length: 1280, dtype: int64


We have a lot of values, but no null values

For each url, check if all the `length` values are the same

In [ ]:
# Group by '_url' and calculate the number of unique values in the 'length' column
length_uniques = dataset.groupby('_url')['length'].nunique(dropna=False)
# Filter the URLs with more than one unique length
multiple_length_urls = length_uniques[length_uniques > 1].index

print('Number of URLs with more than one unique length: ' + str(len(multiple_length_urls)))

Number of URLs with more than one unique length: 0


Since we have a lot of values, we check if every value is sintatically correct.

In [ ]:
# Check that every value is a digit
dataset['length'].dtype

dtype('float64')

In [ ]:
# Get rows where 'length' does not end with '.0'
invalid_rows = dataset[~dataset['length'].astype(str).str.endswith('.0')]
                                
print('Number of invalid lengths: ' + str(len(invalid_rows)))
for index, row in invalid_rows.iterrows():
    print(row['_url'], row['length'])

Number of invalid lengths: 954
tour-de-suisse/1992/stage-4 32200.000000000004
tour-de-suisse/1992/stage-4 32200.000000000004
tour-de-suisse/1992/stage-4 32200.000000000004
tour-de-suisse/1992/stage-4 32200.000000000004
tour-de-suisse/1992/stage-4 32200.000000000004
tour-de-suisse/1992/stage-4 32200.000000000004
tour-de-suisse/1992/stage-4 32200.000000000004
tour-de-suisse/1992/stage-4 32200.000000000004
tour-de-suisse/1992/stage-4 32200.000000000004
tour-de-suisse/1992/stage-4 32200.000000000004
amstel-gold-race/2009/result 258600.00000000003
amstel-gold-race/2009/result 258600.00000000003
amstel-gold-race/2009/result 258600.00000000003
amstel-gold-race/2009/result 258600.00000000003
amstel-gold-race/2009/result 258600.00000000003
amstel-gold-race/2009/result 258600.00000000003
amstel-gold-race/2009/result 258600.00000000003
amstel-gold-race/2009/result 258600.00000000003
amstel-gold-race/2009/result 258600.00000000003
amstel-gold-race/2009/result 258600.00000000003
amstel-gold-race/20

Check the races where the `length` value is small or large, for possible outliers

In [ ]:
# Dataset info, for 'length' column
dataset['length'].describe()

count    589865.000000
mean     166776.180584
std       64545.605664
min        1000.000000
25%      152500.000000
50%      178200.000000
75%      203500.000000
max      338000.000000
Name: length, dtype: float64

The lenghts are between 1000 and 338000 m, so we can say that no extreme values are present.

### Check on 'climb_total' data

Now we consider the `climb_total` column, and check the number of null values and the count the occurrences of each unique value

In [ ]:
print('Total number of null values in climb_total column: ' + str(dataset['climb_total'].isnull().sum())
      + ' (' + str(round(dataset['climb_total'].isnull().sum() / len(dataset) * 100, 2)) + '%)')

print('\nCount occurrences of each value in climb_total column:')
climb_total_counts = dataset['climb_total'].value_counts()
print(climb_total_counts)

Total number of null values in climb_total column: 147045 (24.93%)

Count occurrences of each value in climb_total column:
climb_total
3500.0    3762
2500.0    3261
4000.0    3029
3000.0    2938
5000.0    2592
          ... 
3742.0       9
2525.0       9
1176.0       8
1903.0       5
3128.0       1
Name: count, Length: 2117, dtype: int64


We have a lot of different values, and a lot of null values

For each url, check if all the `climb_total` values are the same


In [ ]:
# Group by '_url' and calculate the number of unique values in the 'climb_total' column
climb_total_uniques = dataset.groupby('_url')['climb_total'].nunique(dropna=False)
# Filter the URLs with more than one unique climb_total
multiple_climb_total_urls = climb_total_uniques[climb_total_uniques > 1].index

print('Number of URLs with more than one unique climb_total: ' + str(len(multiple_climb_total_urls)))

Number of URLs with more than one unique climb_total: 0


Since we have a lot of values, we check if every value is sintatically correct

In [ ]:
# Check that every value is a digit
dataset['climb_total'].dtype

dtype('float64')

In [ ]:
# Get rows where 'climb_total' does not end with '.0'
invalid_rows = dataset[~dataset['climb_total'].astype(str).str.endswith('.0')].dropna(subset=['climb_total'])

print('Number of invalid climb_total: ' + str(len(invalid_rows)))

Number of invalid climb_total: 0


Check the races where the `climb_total` value is small or large, for possible outliers

In [ ]:
# Dataset info, for 'climb_total' column
dataset['climb_total'].describe()

count    442820.000000
mean       2330.469215
std        1375.710722
min           2.000000
25%        1309.000000
50%        2255.000000
75%        3273.000000
max        6974.000000
Name: climb_total, dtype: float64

The climb values are between 2 and 6974 m, so we can say that no extreme values are present.

### Check on 'profile' data

Now we consider the `profile` column, and check the number of null values and the count the occurrences of each unique value

In [ ]:
print('Total number of null values in profile column: ' + str(dataset['profile'].isnull().sum())
      + ' (' + str(round(dataset['profile'].isnull().sum() / len(dataset) * 100, 2)) + '%)')

print('\nCount occurrences of each value in profile column:')
profile_counts = dataset['profile'].value_counts()
print(profile_counts)

Total number of null values in profile column: 148194 (25.12%)

Count occurrences of each value in profile column:
profile
1.0    131344
2.0    128269
5.0     88203
3.0     50844
4.0     43011
Name: count, dtype: int64


We have few different values, but a lot of null values. Also, we see that every value is sintatically correct

For each url, check if all the `profile` values are the same

In [ ]:
# Group by '_url' and calculate the number of unique values in the 'profile' column
profile_uniques = dataset.groupby('_url')['profile'].nunique(dropna=False)
# Filter the URLs with more than one unique profile
multiple_profile_urls = profile_uniques[profile_uniques > 1].index

print('Number of URLs with more than one unique profile: ' + str(len(multiple_profile_urls)))

Number of URLs with more than one unique profile: 0


### Check on 'startlist_quality' data

Now we consider the `startlist_quality` column, and check the number of null values and the count the occurrences of each unique value

In [ ]:
print('Total number of null values in startlist_quality column: ' + str(dataset['startlist_quality'].isnull().sum())
      + ' (' + str(round(dataset['startlist_quality'].isnull().sum() / len(dataset) * 100, 2)) + '%)')

print('\nCount occurrences of each value in startlist_quality column:')
startlist_quality_counts = dataset['startlist_quality'].value_counts()
print(startlist_quality_counts)

Total number of null values in startlist_quality column: 0 (0.0%)

Count occurrences of each value in startlist_quality column:
startlist_quality
971     8279
1812    7807
1872    7317
1612    7255
920     6715
        ... 
455        5
438        5
394        5
228        3
544        3
Name: count, Length: 697, dtype: int64


We have different values, but no null values.

For each url, check if all the `startlist_quality` values are the same

In [ ]:
# Group by '_url' and calculate the number of unique values in the 'startlist_quality' column
startlist_quality_uniques = dataset.groupby('_url')['startlist_quality'].nunique(dropna=False)
# Filter the URLs with more than one unique startlist_quality
multiple_startlist_quality_urls = startlist_quality_uniques[startlist_quality_uniques > 1].index

print('Number of URLs with more than one unique startlist_quality: ' + str(len(multiple_startlist_quality_urls)))

Number of URLs with more than one unique startlist_quality: 0


Since we have a lot of different values, we check if every value is sintatically correct

In [ ]:
# Check that every value is a digit
dataset['startlist_quality'].dtype

dtype('int64')

## Check on 'date' data

Now we consider the `date` column, and check the number of null values and the count the occurrences of each unique value. We do this considerig the merged dataset where we don't have time value of the date

In [ ]:
print('Total number of null values in date column: ' + str(dataset['date'].isnull().sum())
      + ' (' + str(round(dataset['date'].isnull().sum() / len(dataset) * 100, 2)) + '%)')

print('\nCount occurrences of each value in date column:')
date_counts = dataset['date'].value_counts()
print(date_counts)

Total number of null values in date column: 0 (0.0%)

Count occurrences of each value in date column:
date
1998-06-18    407
1987-07-05    404
2006-05-15    379
2006-05-16    377
2023-02-25    376
             ... 
1972-06-23      1
1972-06-21      1
1972-06-17      1
1972-06-22      1
2001-06-02      1
Name: count, Length: 4708, dtype: int64


We have different values, but no null values.

For each url, check if all the `date` values are the same

In [ ]:
# Group by '_url' and calculate the number of unique values in the 'date' column
date_uniques = dataset.groupby('_url_race')['date'].nunique(dropna=False)
# Filter the URLs with more than one unique date
multiple_date_urls = date_uniques[date_uniques > 1].index

print('Number of URLs with more than one unique date (format yyyy-mm-dd, in the merged dataset): ' + str(len(multiple_date_urls)))

Number of URLs with more than one unique date (format yyyy-mm-dd hh:mm:ss, in the races dataset): 4777
Number of URLs with more than one unique date (format yyyy-mm-dd, in the merged dataset): 0


Since we have a lot of different values, we check if every value is sintatically correct

In [ ]:
# Check if there are any 'date' values not in the format yyyy-mm-dd (in the merged dataset)
invalid_rows = dataset[dataset['date'].str.match(r'\d{4}-\d{2}-\d{2}')]
print('Number of invalid dates (format yyyy-mm-dd, in the merged dataset): ' + str(len(invalid_rows)))

Number of invalid dates (format yyyy-mm-dd hh:mm:ss, in the races dataset): 0
Number of invalid dates (format yyyy-mm-dd, in the merged dataset): 0


Check if the year is the same in both the `_url` and the `date`

In [ ]:
# Split _url by / into name, year, and stage
url_split = dataset['_url'].str.split('/', expand=True) # expand=True to return a DataFrame
# Extract the year from the date column (assuming format yyyy-mm-dd hh:mm:ss)
date_year = dataset['date'].str[:4]
# Compare the year in the _url (from the second part of the split) with the year in the date
mismatched_years = dataset[(url_split[1] != date_year)]

# Print the number of rows where the year does not match
print(f"Number of rows where the year in the url does not match the year in the date: {len(mismatched_years)}")


Number of rows where the year in the url does not match the year in the date: 0


### Check on 'position' data

Now we consider the `position` column, and check the number of null values and the count the occurrences of each unique value

In [ ]:
print('Total number of null values in position column: ' + str(dataset['position'].isnull().sum())
      + ' (' + str(round(dataset['position'].isnull().sum() / len(dataset) * 100, 2)) + '%)')

print('\nCount occurrences of each value in position column:')
position_counts = dataset['position'].value_counts()
print(position_counts)

Total number of null values in position column: 0 (0.0%)

Count occurrences of each value in position column:
position
0      5281
1      5275
2      5273
3      5267
4      5255
       ... 
205       5
206       3
207       2
208       2
209       2
Name: count, Length: 210, dtype: int64


We have different values, and no null values.

Since we have different values, we check if every value is sintatically correct

In [ ]:
# Check that every value is a digit
dataset['position'].dtype

dtype('int64')

For each url, check if there are all the `position` values

In [ ]:
# Function to check if the positions are from 0 to the max one after the other
def check_positions(positions):
    return np.array_equal(np.sort(positions), np.arange(positions.max() + 1))

# Apply the function to the dataset
invalid_urls = dataset.groupby('_url')['position'].apply(lambda x: not check_positions(x))

# Stampa gli '_url' che non rispettano la condizione
print('Number of URLs with invalid positions: ' + str(len(invalid_urls[invalid_urls])))


Number of URLs with invalid positions: 0


### Check on 'cyclist_age' data

Now we consider the `cyclist_age` column, and check the number of null values and the count the occurrences of each unique value

In [ ]:
print('Total number of null values in cyclist_age column: ' + str(dataset['cyclist_age'].isnull().sum())
      + ' (' + str(round(dataset['cyclist_age'].isnull().sum() / dataset.shape[0] * 100, 2)) + '%)')

print('\nCount occurrences of each value in cyclist_age column:')
cyclist_age_counts = dataset['cyclist_age'].value_counts()
print(cyclist_age_counts)

Total number of null values in cyclist_age column: 113 (0.02%)

Count occurrences of each value in cyclist_age column:
cyclist_age
27.0    58897
26.0    57921
28.0    56213
25.0    54346
29.0    52616
30.0    46860
24.0    43252
31.0    40827
32.0    35063
23.0    30009
33.0    27972
34.0    21965
35.0    16413
22.0    13531
36.0    12111
37.0     8141
38.0     4703
21.0     3733
39.0     2470
40.0     1127
20.0      842
41.0      307
42.0      218
43.0       96
19.0       90
45.0       16
44.0       11
56.0        1
13.0        1
Name: count, dtype: int64


We have different values, and just a few of null values. Also, we see that every value is sintatically correct

Check the races where the `cyclist_age` value is small or large, for possible outliers

In [ ]:
# Dataset info, for 'cyclist_age' column
dataset['cyclist_age'].describe()

count    589752.000000
mean         28.486208
std           3.855631
min          13.000000
25%          26.000000
50%          28.000000
75%          31.000000
max          56.000000
Name: cyclist_age, dtype: float64

The ages are between 13 and 56 cm, so we can say that no extreme values are present (in the cyclism context).

### Check on 'mostly_tarmac' data

Now we consider the `mostly_tarmac` column, and check the number of null values and the count the occurrences of each unique value

In [ ]:
print('Total number of null values in mostly_tarmac column: ' + str(dataset['mostly_tarmac'].isnull().sum())
      + ' (' + str(round(dataset['mostly_tarmac'].isnull().sum() / len(dataset) * 100, 2)) + '%)')

print('\nCount occurrences of each value in mostly_tarmac column:')
is_tarmac_counts = dataset['mostly_tarmac'].value_counts()
print(is_tarmac_counts)

Total number of null values in is_tarmac column: 0 (0.0%)

Count occurrences of each value in is_tarmac column:
is_tarmac
True     536042
False     53823
Name: count, dtype: int64


We have two different values, and no null values. Also, we see that every value is sintatically correct

### Check on 'cyclist_team' data

Now we consider the `cyclist_team` column, and check the number of null values and the count the occurrences of each unique value

In [ ]:
# Print total number of null values in 'delta' column, and the percentage of null values (float with two decimal digits after the comma)
print('Total number of null values in cyclist_team column: ' + str(dataset['cyclist_team'].isnull().sum())
      + ' (' + str(round(dataset['cyclist_team'].isnull().sum() / dataset.shape[0] * 100, 2)) + '%)')

print('\nCount occurrences of each value in cyclist_team column:')
cyclist_team_counts = dataset['cyclist_team'].value_counts()
print(cyclist_team_counts)

Total number of null values in cyclist_team column: 159161 (26.98%)

Count occurrences of each value in cyclist_team column:
cyclist_team
liberty-seguros-wurth-team-2005     8869
roompot-nederlandse-loterij-2018    8773
chazal-vetta-mbk-1993               8094
kondor-1979                         7895
kazakhstan-2019                     7701
                                    ... 
atala-ofmega-1988                   1259
finland-2016                        1236
south-africa-1993                   1174
denmark-2003                         216
quickstep-innergetic-2009              3
Name: count, Length: 91, dtype: int64


In [ ]:
# Check if every 'cyclist_team' value matches the required format
invalid_teams = dataset[dataset['cyclist_team'].notnull() & ~dataset['cyclist_team'].astype(str).str.match(r'.+-\d{4}')]

print('Number of invalid rows: ' + str(len(invalid_teams)))

Number of invalid rows: 0


### Check on 'delta' data

Now we consider the `delta` column, and check the number of null values and the count the occurrences of each unique value

In [ ]:
print('Total number of null values in delta column: ' + str(dataset['delta'].isnull().sum())
        + ' (' + str(round(dataset['delta'].isnull().sum() / dataset.shape[0] * 100, 2)) + '%)')

print('\nCount occurrences of each value in delta column:')
delta_counts = dataset['delta'].value_counts()
print(delta_counts)

Total number of null values in delta column: 0 (0.0%)

Count occurrences of each value in delta column:
delta
0.0       120546
2.0         3700
5.0         3353
3.0         3178
4.0         3036
           ...  
2453.0         1
2357.0         1
2889.0         1
2718.0         1
2527.0         1
Name: count, Length: 2836, dtype: int64


We have lots of different values, but no null values.

Since we have a lot of different values, we check if every value is sintatically correct.

In [ ]:
# Check that every value is a digit
dataset['delta'].dtype

dtype('float64')

In [ ]:
# Check if there are any 'delta' values that do not end with '.0'
invalid_rows = dataset[~dataset['delta'].astype(str).str.endswith('.0')]

print('Number of invalid deltas: ' + str(len(invalid_rows)))

Number of invalid deltas: 0


In [ ]:
# Check if there are any 'delta' values that are negative
invalid_rows = dataset[dataset['delta'] < 0]

print('Number of negative deltas: ' + str(len(invalid_rows)))
for index, row in invalid_rows.iterrows():
    print(row['_url'], row['delta'])

Number of negative deltas: 86
vuelta-a-espana/1992/stage-19 -2635.0
vuelta-a-espana/1992/stage-19 -2638.0
vuelta-a-espana/1992/stage-19 -2541.0
vuelta-a-espana/1992/stage-19 -2542.0
vuelta-a-espana/1992/stage-19 -2545.0
vuelta-a-espana/1992/stage-19 -2546.0
vuelta-a-espana/1992/stage-19 -2550.0
vuelta-a-espana/1992/stage-19 -2560.0
vuelta-a-espana/1992/stage-19 -2564.0
vuelta-a-espana/1992/stage-19 -2567.0
vuelta-a-espana/1992/stage-19 -2567.0
vuelta-a-espana/1992/stage-19 -2574.0
vuelta-a-espana/1992/stage-19 -2469.0
vuelta-a-espana/1992/stage-19 -2473.0
vuelta-a-espana/1992/stage-19 -2475.0
vuelta-a-espana/1992/stage-19 -2477.0
vuelta-a-espana/1992/stage-19 -2479.0
vuelta-a-espana/1992/stage-19 -2481.0
vuelta-a-espana/1992/stage-19 -2482.0
vuelta-a-espana/1992/stage-19 -2485.0
vuelta-a-espana/1992/stage-19 -2486.0
vuelta-a-espana/1992/stage-19 -2486.0
vuelta-a-espana/1992/stage-19 -2487.0
vuelta-a-espana/1992/stage-19 -2500.0
vuelta-a-espana/1992/stage-19 -2504.0
vuelta-a-espana/1992

Check if following the `positon` order, the delta is ordered too

In [ ]:
# Grouped by '_url', check if the 'delta' value is greater or equal than the previous one (the one in the previous row, except for the first one)
invalid_urls = dataset.groupby('_url')['delta'].apply(lambda x: (x.shift() > x).any())

# Print the number of URLs with invalid 'delta' values
print('Number of URLs with invalid deltas: ' + str(len(invalid_urls[invalid_urls])))
for url in invalid_urls[invalid_urls].index:
    print(url)

Number of URLs with invalid deltas: 346
dauphine/2006/stage-2
dauphine/2007/stage-7
dauphine/2008/stage-2
dauphine/2008/stage-3
dauphine/2008/stage-6
dauphine/2008/stage-7
dauphine/2009/stage-4
dauphine/2009/stage-6
dauphine/2011/stage-2
dauphine/2015/stage-2
dauphine/2016/stage-4
dauphine/2017/stage-1
dauphine/2022/stage-1
dauphine/2022/stage-2
dauphine/2023/stage-1
dauphine/2023/stage-3
dauphine/2023/stage-5
dwars-door-vlaanderen/2003/result
e3-harelbeke/1976/result
e3-harelbeke/2023/result
giro-d-italia/1991/stage-10
giro-d-italia/1992/stage-3
giro-d-italia/1994/stage-8
giro-d-italia/1995/stage-14
giro-d-italia/1995/stage-17
giro-d-italia/1995/stage-5
giro-d-italia/1995/stage-8
giro-d-italia/1997/stage-13
giro-d-italia/1997/stage-3
giro-d-italia/1999/stage-7
giro-d-italia/2002/stage-2
giro-d-italia/2002/stage-3
giro-d-italia/2002/stage-4
giro-d-italia/2003/stage-1
giro-d-italia/2005/prologue
giro-d-italia/2005/stage-1
giro-d-italia/2005/stage-15
giro-d-italia/2005/stage-2
giro-d-ita

### Check on 'year' data

Now we consider the `year` column, and check the number of null values and the count the occurrences of each unique value

In [20]:
# Print total number of null values in 'delta' column, and the percentage of null values (float with two decimal digits after the comma)
print('Total number of null values in year column: ' + str(dataset['year'].isnull().sum())
      + ' (' + str(round(dataset['year'].isnull().sum() / dataset.shape[0] * 100, 2)) + '%)')

print('\nCount occurrences of each value in year column:')
cyclist_team_counts = dataset['year'].value_counts()
print(cyclist_team_counts)

KeyError: 'year'

### Check on 'stage' data

Now we consider the `stage` column, and check the number of null values and the count the occurrences of each unique value

In [19]:
# Print total number of null values in 'delta' column, and the percentage of null values (float with two decimal digits after the comma)
print('Total number of null values in stage column: ' + str(dataset['stage'].isnull().sum())
      + ' (' + str(round(dataset['cyclist_team'].isnull().sum() / dataset.shape[0] * 100, 2)) + '%)')

print('\nCount occurrences of each value in stage column:')
cyclist_team_counts = dataset['stage'].value_counts()
print(cyclist_team_counts)

KeyError: 'stage'